### PR TITLE
Configmgr 2.0 Initial Baseline

### PR DESCRIPTION
**Pull request by garonsky**
```
- ConfigMgr 2.0 Initial baseline
```

Refer to: https://github.com/hpcc-systems/HPCC-Platform/pull/7531

```
    - Changes include
    - Code refactoring to remove Dojo UI prototype code, removal of unneeded files, and compaction of code.
    - Comments in previous commit incorporated
    - Specifically:  
        - tighter more compact code
        - most comments removed
        - prototyped code removed
        - standard use of UNIMPLEMENTED macro
        - confirmed standard use of QT environment variable for builds
        - removed static variable for string arrays
    - Added copyright
```

Signed-off-by: Gleb Aronsky gleb.aronsky@lexisnexis.com

### CODE REVIEW

**Comment by garonsky** [comment]
@richardkchapman please review

File: configuration/configurator/BuildSet.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+    for (int idx = 0; idx < nLength; idx++)
+    {
+        buildSetArray.append(this->getBuildSetComponentName(idx));
+    }
+}
+
+void CBuildSetManager::getBuildSetServices(StringArray& buildSetArray) const
+{
+    int nLength = this->getBuildSetServiceCount();
```
Comment (richardkchapman) [inline configuration/configurator/BuildSet.cpp]: Why all the "this->" ?

File: cmake_modules/commonSetup.cmake [hunk 2]
```diff
@@ -91,6 +91,17 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
...
   option(USE_REDIS "Enable Redis support" ON)
+  option(MAKE_CONFIGURATOR "Build Configurator" ON)
+  option(CONFIGURATOR_LIB "Build Configurator static library (.a)" OFF)
+  option(CONFIGURATOR_QT_UI "Build Configurator with QT based UI" OFF)
+
+  if (CONFIGURATOR_QT_UI OR CONFIGURATOR_LIB )
+        set( MAKE_CONFIGURATOR ON )
+  endif()
+
+  if ( (CONFIGURATOR_QT_UI AND CONFIGURATOR_LIB) )
```
Comment (richardkchapman) [inline cmake_modules/commonSetup.cmake]: I don't really understand the logic here. Why would we want to forbid building both (if other components in the build need the static lib, then presumably it needs to be built. If they don't, then why have the option at all?)

File: configuration/configurator/BuildSet.cpp [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    StringBuffer xpath;
+
+    if (m_buildSetTree.get() != NULL)
+        return false;
+
+    try
+    {
+        m_buildSetTree.set(createPTreeFromXMLFile(m_buildSetPath.str()));
+    }
+    catch(...)
```
Comment (richardkchapman) [inline configuration/configurator/BuildSet.cpp]: You will leak the IException that createPTreeFromXMLFile throws.

Also, should you be logging something about exactly why you could not load the buildset?

File: configuration/configurator/ConfigFileComponentUtils.hpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+class CConfigFileComponentUtils : public CInterface
+{
+public:
+
+    IMPLEMENT_IINTERFACE
+
+    CConfigFileComponentUtils();
+    virtual ~CConfigFileComponentUtils();
+
+    void getAvailableComponets(StringArray& compArray) const;
```
Comment (richardkchapman) [inline configuration/configurator/ConfigFileComponentUtils.hpp]: typo: missing n in componeNts

File: configuration/configurator/ConfigFileUtils.hpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+#ifndef _CONFIG_FILE_HPP_
+#define _CONFIG_FILE_HPP_
+
+#include "jiface.hpp"
+#include "jmutex.hpp"
+#include "jutil.hpp"
+#include "ConfigFileUtilsObservable.hpp"
+
+static const char* DEFAULT_CONFIGURATION_PATH("/etc/HPCCSystems/source/");
```
Comment (richardkchapman) [inline configuration/configurator/ConfigFileUtils.hpp]: Should do this via the generated build-config.h file

File: configuration/configurator/ConfiguratorAPI.cpp [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+
+    return CConfigSchemaHelper::getInstance()->getElementArraySize(pXPath);
+}
+
+int getNumberOfTables()
+{
+    return CConfigSchemaHelper::getInstance()->getNumberOfTables();
+}
+
+const char* getServiceName(int idx, char *pName)
```
Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.cpp]: Use a StringBuffer - otherwise you are opening up an avenue for buffer overruns

Comment (richardkchapman) [inline configuration/configurator/ConfiguratorAPI.cpp]: (if there's a reason you CAN'T use a StringBuffer, then you need to pass in the target buffer size and check that the length you are copying does not exceed it).

File: configuration/xsds_xmls/environment2.xml [hunk 7]
```diff
@@ -0,0 +1,1123 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (richardkchapman) [inline configuration/xsds_xmls/environment2.xml]: Is this file supposed to be checked in?

**Comment by richardkchapman** [comment]
A file (README.md would be perfect) with a quick description of how the code is structured and what the function of the various bits of it are would be useful both for reviewing and subsequent maintenance.

**Comment by richardkchapman** [comment]
@garonsky see comments

**Pull request closed without merging**